### PR TITLE
[hannk] Allow aliasing of Reshape tensors

### PR DESCRIPTION
**Pull request by steven-johnson**
Previously we didn't allow this because aliased tensors had to have the same rank, which is ~never the case for Reshape. Aliasing for Reshape is a huge win because it essentially becomes a no-op rather than a memcpy. Running against standard set of models shows no regression in differences vs. tflite.

### CODE REVIEW

File: apps/hannk/interpreter/ops.cpp [hunk 1]
```diff
@@ -1400,12 +1400,16 @@ void ReshapeOp::execute() {
         assert(new_shape.at(d) == output_buf.dim(d).extent());
     }
 
-    assert(input_buf.number_of_elements() == output_buf.number_of_elements());
+    // NOTE: HCHECK, not assert, to ensure a runtime failure to match TFLite
```
Comment (dsharletg) [inline apps/hannk/interpreter/ops.cpp:1403]: We should still be asserting here. This check should be in the parser (i.e. not running every iteration).

**Comment by steven-johnson** [comment]
Ehh, I think this needs more work, actually: it doesn't account for the case of the RHS of the set_alias_of() call having a storage_offset.

File: apps/hannk/interpreter/tensor.cpp [hunk 2]
```diff
@@ -192,13 +222,20 @@ void Tensor::set_alias_of(const TensorPtr &t, const SmallVector<int, max_rank> &
...
     auto &shared_buffer = storage_->buffer;
     assert(shared_buffer.type().bytes() == type().bytes());
-    assert(shared_buffer.dimensions() == (int)offset_bounds.size());
     assert(!shared_buffer.data());
 
     // Check that the storage is big enough for this buffer.
-    for (int i = 0; i < shared_buffer.dimensions(); i++) {
-        assert(offset_bounds[i].min >= shared_buffer.dim(i).min());
-        assert(offset_bounds[i].max <= shared_buffer.dim(i).max());
+    if (this->rank() == t->rank()) {
```
Comment (dsharletg) [inline apps/hannk/interpreter/tensor.cpp:228]: I think this is a little sketchy. I think reshapes of the same rank are legal, where each dimension has a different extent, but the total number of elements is the same. I think this either needs to be a separate function, or a parameter indicating the aliasing should share "flat" indices or "rank-dimensional" indices.

File: apps/hannk/interpreter/ops.cpp (hunk 1 above)
Comment (steven-johnson) [reply apps/hannk/interpreter/ops.cpp:1403]: The dynamic-tensor case still needs a runtime check.

File: apps/hannk/interpreter/tensor.cpp (hunk 2 above)
Comment (steven-johnson) [reply apps/hannk/interpreter/tensor.cpp:228]: Originally I just had a flag inside the Tensor indicating whether the aliasing was for a reshape or not, but that felt clumsy -- I didn't think about the same-rank, different-extent case though. I'll revisit this.

**Comment by steven-johnson** [comment]
PTAL. (Added some drive-by debugging changes to lower_tflite_lstm() and Tensor::dump as well.)

**Review by dsharletg — APPROVED** [review]

**Pull request merged**